File: 2023/UICPC DIV1/UICPC Round #06/solutions & data/drivinglanes/submissions/accepted/drivinglanes-db.cc
#include <bits/stdc++.h>
using namespace std;

vector<vector<long long> > DP;
int n, m, k, r;
vector<int> L;
vector<pair<int,int> > C;

long long f(int seg, int l){
  if(seg == n) return (l == 0 ? 0 : LLONG_MAX/2);
  long long& x = DP[seg][l];
  if(x >= 0) return x;
  
  x = LLONG_MAX/2;
  for(int i=0;i<m;i++)
    if( abs(l-i) * k <= L[seg] ){
      long long curr = f(seg+1, i) + L[seg] +
	abs(l-i) * r + C[seg].first + C[seg].second * (i+1);
      x = min(x, curr);
    }
  
  return x;
}

int main(){
  cin >> n >> m >> k >> r;
  
  L.resize(n);
  for(auto& x : L) cin >> x;
  
  C.resize(n-1);
  for(auto& x : C) cin >> x.first >> x.second;
  C.emplace_back(0,0); // Make life easier...
  
  DP.assign(n+1, vector<long long>(m+1, -1));
  
  cout << f(0,0) << endl;
}

File: 2023/UICPC DIV1/UICPC Round #06/solutions & data/drivinglanes/submissions/accepted/drivinglanes_db.java
import java.util.*;

public class drivinglanes_db {
  public static void main(String[] args) {
    new drivinglanes_db().run();
  }

  final long oo = Long.MAX_VALUE/2;
  long[][] DP;
  long[] L;
  long[] C;
  long[] S;

  int n,m,k,r;
  
  void run() {
    Scanner in = new Scanner(System.in);

    n = in.nextInt(); m = in.nextInt();
    k = in.nextInt(); r = in.nextInt();

    L = new long[n];
    for(int i=0;i<n;i++)
      L[i] = in.nextLong();
    
    C = new long[n];
    S = new long[n];
    for(int i=0;i<n-1;i++){
      S[i] = in.nextLong();
      C[i] = in.nextLong();
    }

    DP = new long[n+1][m+1];

    for(int i=0;i<=n;i++)
      for(int j=0;j<=m;j++)
	DP[i][j] = -1;

    System.out.println(f(0,0));
  }

  long f(int seg, int l){
    if(seg == n) return (l == 0 ? 0 : oo);
    long x = DP[seg][l];
    if(x >= 0) return x;
    
    x = oo;
    for(int i=0;i<m;i++)
      if( Math.abs(l-i) * k <= L[seg] ){
	long curr = f(seg+1, i) + L[seg] +
	  Math.abs(l-i) * r + S[seg] + C[seg] * (i+1);
	x = Math.min(x, curr);
      }
    
    return DP[seg][l] = x;
  }
}
